### ISSUE TITLE
amazon: improve version constraint construction for ALASKERNEL-x.x-* advisories to reduce false positive matches

### ISSUE THREAD

**Issue by westonsteimel** [anchore/vunnel#265]
**What happened**:

Amazon linux advisories like [ALASKERNEL-5.10-2022-005](https://alas.aws.amazon.com/AL2/ALASKERNEL-5.10-2022-005.html) should only apply to that specific kernel series (5.10.x), because they create an advisory per kernel line in these cases:

So ALASKERNEL-5.10-2022-005 patches [CVE-2021-3753](https://alas.aws.amazon.com/cve/html/CVE-2021-3753.html) which shows:

<img width="1653" alt="image" src="https://github.com/anchore/vunnel/assets/1593939/f67fdbc7-1006-4773-9ed1-eed7429e98ab">

Right now since we construct the affected version constraint as `< 5.10.62-55.141.amzn2` for ALASKERNEL-5.10-2022-005, it means that even a 4.x series kernel with the patch from [ALAS-2021-1704](https://alas.aws.amazon.com/AL2/ALAS-2021-1704.html) or a 5.4 series kernel with the patch from [ALASKERNEL-5.4-2022-007](https://alas.aws.amazon.com/AL2/ALASKERNEL-5.4-2022-007.html).  In fact, it's even worse for a 4.x kernel since it'll potentially report all 3.  After looking at the data I think it should be correct to build a version constraint something like `>= 5.10, < 5.10.62-55.141.amzn2` for these cases.  This should lead to correctly interpreted results for these cases.

There are some other instances where AWS issues multiple advisories across several lines of a product; however, in all of the other cases the package has a distinct name so shouldn't require this workaround, but the kernel packages do not have this property.

**What you expected to happen**:

Only `ALAS-2021-1704` should be reported for a vulnerable 4.x kernel line, and only `ALASKERNEL-5.4-2022-007` should be reported for a vulnerable 5.4.x kernel line.

**How to reproduce it (as minimally and precisely as possible)**:

`dcoker run --rm anchore/grype:v0.65.2 docker.io/anchore/test_images:vulnerabilities-amazonlinux-2-5c26ce9@sha256:cf742eca189b02902a0a7926ac3fbb423e799937bf4358b0d2acc6cc36ab82aa`

**Anything else we need to know?**:

**Environment**:
- Output of `vunnel version`:
- OS (e.g: `cat /etc/os-release` or similar):

**Comment by westonsteimel** [comment]
Hmm, actually since the vunnel schema only outputs when a fix happened, I don't think this can be handled here but might need to be deferred to grype-db

**Comment by westonsteimel** [comment]
Or should it be some sort of change to the schema to add in an optional LowerBound property to the FixeIn entry?

**Comment by westonsteimel** [comment]
I created https://github.com/anchore/vunnel/issues/266 to start tracking what we might want to consider when designing future vunnel major version schemas

**Comment by Dror-Madar** [comment]
Measurement from downstream, in case it's useful: **this is still live in the shipped database**, and the "kernel is the only affected package" assumption doesn't hold on current data.

I hit this while measuring a second matcher against Grype on Amazon Linux, and went back to find whether it had been handled — [the resolution here](https://github.com/anchore/vunnel/issues/265#issuecomment-1691733271) was to work around it in grype-db rather than vunnel. As far as I can tell that workaround never shipped.

**Shipped DB v6.1.9 (built 2026-09-17):**

```
amazonlinux affected-package rows:              60,991
...carrying a lower bound of any kind:               0
```

Queried straight out of `vulnerability.db`:

```sql
SELECT vh.name, p.name, b.value
FROM affected_package_handles aph
JOIN vulnerability_handles vh ON vh.id = aph.vulnerability_id
JOIN packages p ON p.id = aph.package_id
JOIN blobs b ON b.id = aph.blob_id
JOIN operating_systems os ON os.id = aph.operating_system_id
WHERE os.name = 'amazonlinux';
```

```
ALAS2KERNEL-5.10-2022-001 | bpftool | < 5.10.35-31.135.amzn2
ALAS2KERNEL-5.4-2022-001  | bpftool | < 5.4.105-48.177.amzn2
ALAS2KERNEL-5.15-2022-001 | bpftool | < 5.15.43-20.123.amzn2
```

So an AL2 host on `4.14.355-277.647.amzn2` is still matched by all three, exactly as originally described.

### The "distinct name" assumption

The issue reasons:

> There are some other instances where AWS issues multiple advisories across several lines of a product; however, in all of the other cases the package has a distinct name so shouldn't require this workaround, but the kernel packages do not have this property.

That isn't true of the current data. Grouping by `(os, package)` and keeping only packages that appear under more than one ALAS product *with different version lines* gives **138 package/OS keys across 22 product groups**, 10,874 affected-package rows. `ALAS2PHP8.0`, `ALAS2PHP8.1` and `ALAS2PHP8.2` all ship a package named exactly `php`:

```
23 pkgs  ALAS2 + ALAS2PHP8.0 + ALAS2PHP8.1 + ALAS2PHP8.2      (php, php-cli, php-common, …)
14 pkgs  ALAS2 + ALAS2RUBY2.6 + ALAS2RUBY3.0                  (ruby, ruby-libs, …)
12 pkgs  ALAS2 + ALAS2KERNEL-5.4/5.10/5.15                    (kernel, perf, bpftool, …)
10 pkgs  ALAS2 + ALAS2POSTGRESQL11/12/13/14                   (postgresql, libpq, …)
10 pkgs  ALAS2R3.4 + ALAS2R4                                  (R, R-core, …)
 6 pkgs  ALAS2 + ALAS2TOMCAT8.5 + ALAS2TOMCAT9                (tomcat, …)
 …       golang, mariadb, squid, unbound, memcached, dnsmasq
```

### One caveat on the suggested constraint

`>= 5.10, < 5.10.62-55.141.amzn2` is the right shape, but deriving the lower bound from the advisory id alone is not safe on its own. **563 rows** carry a stream in the id and a fix version outside it, because one advisory ships packages that don't track the line it's named for:

```
ALAS2KERNEL-5.10  stream 5.10  fix 1.0-0.amzn2          (kernel-livepatch-*)   3,523 rows
ALAS2KERNEL-5.15  stream 5.15  fix 1.0-0.amzn2                                 3,007 rows
ALAS2CORRETTO8    stream 8     fix 1.8.0_312.b07-2.amzn2                         278 rows
ALAS2RUBY2.6      stream 2.6   fix 1.4.1-129.amzn2.0.1  (rubygems)
```

A blind `>= 8` on the Corretto rows produces a constraint that matches nothing, turning a false positive into a false negative. Gating on "the advisory's own fixed version is itself inside the stream named by the id" drops all of those cleanly.

Happy to open a PR against grype-db if that's still the preferred place for it — just wanted the measurement on the record either way, since we're carrying a local filter for it in the meantime.
